proposals/0534-metro-package-exports-support.md

```diff
@@ -0,0 +1,455 @@
+
+This can be seen in the defaults for Metro's [`resolverMainFields` option](https://facebook.github.io/metro/docs/configuration/#resolvermainfields) (applied by React Native CLI). As this is a static list of top-level fields, the resolver will invariantly prioritise `"react-native"` over `"browser"`:
+
+```js
+resolverMainFields: ['react-native', 'browser', 'main'],
+```
+
+We believe this default is non-ideal as it inhibits packages using the `"browser"` field spec from targeting react-native-web — instead working only for React Native-specific `.web.js` extensions. We believe the majority of non-React Native community package authors (in addition to React app developers) expect `"browser"` to used on web.
+
+This default is directly misaligned with react-native-web via Webpack  ([docs](https://webpack.js.org/guides/package-exports/#target-environment)), which is used by a majority of users.
+
+**Proposed**:
+
+- **Breaking**: We will implement the `"browser"` condition name — changed from the top-level field to be selected when the React Native platform is `'web'`.
+- **Breaking**: We will introduce a `"react-native"` condition name — changed from the top-level field to represent a native React Native JS runtime environment (and not web).
+- Metro will not provide an implementation for other community conditions.
+
+Combined, this is the only breaking change we foresee in Metro that will immediately impact app developers, and which we are not deferring via strict mode opt-in. It will align behaviour for existing users of react-native-web via Webpack.
+
+| Condition name | Description | Node.js | Webpack | Metro (proposed) |
+| - | - | - | - | - | - |
+| `"browser"` | Any web browser environment | - | Matched when `platform === 'web'` | Matched when `platform === 'web'` |
+| `"react-native"` [New] | Any React Native runtime environment (non-browser) | - | Matched when `platform !== 'web'` | Matched when `platform !== 'web'` |
+| All other | | (unimplemented) | Various supported, e.g. `"deno"`, `"worker"`, `"electron"` | (unimplemented) |
+| `"default"`| Makes no assumption about JS environment | Always matched | Always matched | Always matched |
+
+Note: Now that `"react-native"` and `"browser"` will not overlap, packages intending to support both native and web environments but which want to expose alternative `"exports"` targeting *React Native **APIs*** should continue to use the `"default"` condition and use a [subpath export](#subpath-exports) instead: e.g. `import { reactNativeExport } from 'some-pkg/react-native'`.
+
+### Conditional exports: User conditions
```

Comment:
whatever happens here needs to be accessible to the custom env (again, #509) - maybe the env should get export conditions from Metro config rather than having its own?

Reply:
Sounds ideal. I'd be keen if this was explicitly configured, e.g. specifying a `metroConfigPath` in `ReactNativeEnv` as a configuration point — especially if calling Metro is the solution for [aligning nonstrict exports handling in Jest (your other comment)](https://github.com/react-native-community/discussions-and-proposals/pull/534/files#r997999005).

Otherwise defining user conditions (default = `['react-native']`) in both configs as part of the project template may be adequate.